multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,307 @@
+// computes the label identities added to and deleted from the cluster, and exports them to the
+// leader cluster for further processing.
+type (
+	LabelIdentityReconciler struct {
+		client.Client
+		Scheme           *runtime.Scheme
+		commonAreaGetter RemoteCommonAreaGetter
+		remoteCommonArea commonarea.RemoteCommonArea
+		// labelMutex prevents concurrent access to labelToPodsCache and podLabelCache
+		labelMutex sync.RWMutex
+		// resourceExportMutex prevents concurrent updates to the cluster label identity ResourceExport
+		resourceExportMutex sync.Mutex
+		// labelToPodsCache stores mapping from label identities to Pods that have this label identity
+		labelToPodsCache map[string]sets.String
+		// podLabelCache stores mapping from Pods to their label identities
+		podLabelCache  map[string]string
+		localClusterID string
+	}
+)
+
+func NewLabelIdentityReconciler(
+	client client.Client,
+	scheme *runtime.Scheme,
+	commonAreaGetter RemoteCommonAreaGetter) *LabelIdentityReconciler {
+	return &LabelIdentityReconciler{
+		Client:           client,
+		Scheme:           scheme,
+		commonAreaGetter: commonAreaGetter,
+		labelToPodsCache: map[string]sets.String{},
+		podLabelCache:    map[string]string{},
+	}
+}
+
+// +kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Pod/Namespace for label identity", "pod/ns", req.NamespacedName)
+	var commonArea commonarea.RemoteCommonArea
+	var err error
+	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			return ctrl.Result{}, r.onPodDelete(ctx, req.NamespacedName.String())
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	nsLabels, podLabels := ns.Labels, pod.Labels
+	return ctrl.Result{}, r.onPodCreateOrUpdate(ctx, req.NamespacedName.String(), getNormalizedLabel(nsLabels, podLabels))
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *LabelIdentityReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Ignore status update event via GenerationChangedPredicate
+	instance := predicate.Or(predicate.GenerationChangedPredicate{}, predicate.LabelChangedPredicate{})
```

Comment:
I think `LabelChangedPredicate` does not cover Pod ADD events?

Reply:
My comment is about `GenerationChangedPredicate`. Actually both GenerationChangedPredicate and LabelChangedPredicate are for update event. ADD event will be handled anyway.
My point is there is no generation field for Pod and Namespace which means it actually will stop queue update event for Pod or Namespace. You code works only because you used `Or`. You can check the comment of `GenerationChangedPredicate`:
```
// * The assumption that the Generation is incremented only on writing to the spec does not hold for all APIs.
```

Reply:
Thanks for pointing this out. Fixed by removing `GenerationChangedPredicate`.